File: matlab/associateBackgrounds.m
function [BG]=associateBackgrounds(ROI_N,FULL_BG)

if isempty(FULL_BG)
    disp('Error, there are no backgrounds in these files');
    return; 
end
ind=1:length(FULL_BG.Time);

% +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
% rename elements of array of struct
% https://blogs.mathworks.com/loren/2010/05/13/rename-a-field-in-a-structure-array/
BG=ROI_N.IMAGE;
f=fieldnames(BG);
v=struct2cell(BG);
f{strmatch('IM',f,'exact')}='BG';
BG = cell2struct(v,f);
%--------------------------------------------------------------------------


for i=1:length(ROI_N.IMAGE)
    if length(ind)==1
        jj=ind;
    else
        Timecur=ROI_N.Time(i);
        %Times=ROI_N.Times(ind);
        jj=interp1(FULL_BG.Time,ind,Timecur,'nearest');
    end
    
    BG(i).BG(:,:)=...
        FULL_BG.IMAGE(jj).BG(ROI_N.StartX(i)+1:ROI_N.EndX(i)+1,...
        ROI_N.StartY(i)+1:ROI_N.EndY(i)+1);
end

File: python/associateBackgrounds.py
import numpy as np
from scipy.interpolate import interp1d

def associateBackgrounds(ROI_N,FULL_BG):

    
    if len(FULL_BG['Time'][0,0][0,:]) == 0:
        print('Error, there are no backgrounds in these files')
        
    ind=np.mgrid[0:len(FULL_BG['Time'][0,0][0,:])]

    #--------------------------------------------------------------------------
    BG = [dict() for x in range(len(ROI_N['IMAGE'][0,0][0,:]))]

    for i in range(len(ROI_N['IMAGE'][0,0][0,:])):
       if len(ind)==1:
           jj=int(ind)
       else:
          Timecur=ROI_N['Time'][0,0][i,0]
          #https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html
          f=interp1d(FULL_BG['Time'][0,0][0,:],ind,kind='nearest',fill_value='extrapolate')
          jj=int(f(Timecur))
       
       
       xl=ROI_N['StartX'][0,0][0,i]
       xu=ROI_N['EndX'][0,0][0,i]+1
       yl=ROI_N['StartY'][0,0][0,i]
       yu=ROI_N['EndY'][0,0][0,i]+1
       BG[i]['BG'] = FULL_BG[0,0]['IMAGE'][0,0][0,jj]['BG'][xl:xu,yl:yu]
       
    
    
    return BG